### PR TITLE
Warn when setting a variable that shadow an option

### LINKED ISSUE #4730: Global variables should not shadow options
`opam var --global jobs=42` sets the global variable `jobs` which creates a strange state for opam in that the `jobs` variable is now `42` but opam will only schedule up to the _previous_ value of `jobs`.

The correct course of action here is `opam option jobs=42` - `opam var` should not permit shadowing the options.

### PR DESCRIPTION
**Pull request by rjbou**
ask user what they want to do
fix #4730

### CODE REVIEW

**Review by kit-ty-kate — APPROVED** [review]

File: tests/reftests/var-option.test [hunk 1]
```diff
@@ -428,8 +426,8 @@ Removed variable dolore in global configuration
 ### opam option eval-variables
 []
 ### opam option global-variables
-[[foo "global-foo" "Set through 'opam var'"] [bin "global-bin" "Set through 'opam var'"] [os-version "dolor" "Set through 'opam var'"] [os-family "ipsum" "Set through 'opam var'"] [os-distribution "lorem" "Set through 'opam var'"] [os "linux" "Set through 'opam var'"] [opam-version "68.79" "Set through 'opam var'"] [make "make" "Set through 'opam var'"] [jobs "7" "Set through 'opam var'"] [arch "x86_64" "Set through 'opam var'"] [sys-ocaml-version "4.08.0" "Set through 'opam var'"]]
-### # Check uniable operations
+[[foo "global-foo" "Set through 'opam var'"] [bin "global-bin" "Set through 'opam var'"] [os-version "dolor" "Set through 'opam var'"] [os-family "ipsum" "Set through 'opam var'"] [os-distribution "lorem" "Set through 'opam var'"] [os "linux" "Set through 'opam var'"] [make "make" "Set through 'opam var'"] [arch "x86_64" "Set through 'opam var'"] [sys-ocaml-version "4.08.0" "Set through 'opam var'"]]
+### # Check unable operations
```
Comment (dra27) [inline tests/reftests/var-option.test:430]: Exceedingly minor, but "unavailable" or "disallowed"?

File: src/client/opamConfigCommand.ml [hunk 2]
```diff
@@ -779,6 +779,20 @@ type ('var,'config) var_confset =
...
+    List.exists (fun (name, _) -> String.equal svar name) OpamFile.Config.fields
+  in
+  let switch_option =
+    List.exists (fun (name, _) -> String.equal svar name)
+      OpamFile.Switch_config.fields
+  in
+  if global_option || switch_option then
+    (OpamConsole.msg
+       "You are setting a variable that have the same name than a %s option.\n\
+        Maybe you want to use 'opam option'\n"
```
Comment (dra27) [inline src/client/opamConfigCommand.ml:792]:
```suggestion
       "You are setting a variable that has the same name as a %s option.\n\
        Maybe you want to use 'opam option' instead\n"
```

File: src/client/opamConfigCommand.ml [hunk 3]
```diff
@@ -779,6 +779,20 @@ type ('var,'config) var_confset =
...
 let set_var svar value conf =
+  let global_option =
+    List.exists (fun (name, _) -> String.equal svar name) OpamFile.Config.fields
+  in
+  let switch_option =
+    List.exists (fun (name, _) -> String.equal svar name)
+      OpamFile.Switch_config.fields
+  in
+  if global_option || switch_option then
+    (OpamConsole.msg
```
Comment (dra27) [inline src/client/opamConfigCommand.ml:790]: Slightly concerned about compatibility - although it's almost guaranteed to be the case that setting, say, `jobs` with `opam var` was a mistake, there could be situations where this is correct. Can the question be guarded either for when we're in interactive mode or for when we're in the default or 2.2 CLI?

**Review by dra27 — COMMENTED** [review]
Wondering if this should be a warning, rather than a prompt?

File: tests/reftests/var-option.test (hunk 1 above)
Comment (rjbou) [reply tests/reftests/var-option.test:430]:
```suggestion
### # Check disallowed operations
```

File: src/client/opamConfigCommand.ml (hunk 3 above)
Comment (dra27) [reply src/client/opamConfigCommand.ml:790]: Fresh thought, coming back to this - how about displaying a warning and advertising the command to revert to it (`opam var set FOO=`) but then setting it without any prompt. It's almost certainly a mistake on the user's part, so we tell them how to fix it, but we still do as we were told (if under protest!)

**Comment by rjbou** [comment]
Updated!

File: src/client/opamConfigCommand.ml [hunk 4]
```diff
@@ -936,6 +936,21 @@ type ('var,'config) var_confset =
...
+  let switch_option =
+    List.exists (fun (name, _) -> String.equal svar name)
+      OpamFile.Switch_config.fields
+  in
+  if global_option || switch_option then
+    (let scope = if global_option then "global" else "switch" in
+     OpamConsole.warning
+       "You are setting a variable that has the same name as a %s option.\n\
+        Did you mean to use `opam option` instead? You can revert this variable's \
+        setting with: 'opam var %s= --%s'\n"
```
Comment (kit-ty-kate) [inline src/client/opamConfigCommand.ml:952]:
```suggestion
        Did you mean to use `opam option` instead? You can revert this variable \
        using: 'opam var %s= --%s'\n"
```

**Review by kit-ty-kate — APPROVED** [review]

**Pull request merged**